### DISCUSSION TITLE
Stuck trying to make my own calibrated colors working

### DISCUSSION THREAD

**Discussion by Rogue75** [pybricks/support#1334] [Q&A] [answered]
To preface, I'm new to Python and started with the tutorial by EggyBricks https://www.youtube.com/watch?v=Jmh4NLxJ_8c
I have the City Hub + Mindstorms Color Sensor ( LEGO Technic Color Sensor by LEGO Education (45605) )

I have this color code from the docs input:

```python
sensor = ColorSensor(Port.B) 

Color.YELLOW = Color(h=50, s=20, v=25)
Color.GREEN = Color(h=110, s=30, v=19)
Color.RED = Color(h=354, s=45, v=15)
Color.WHITE = Color(h=195, s=20, v=24)
Color.GRAY = Color(h=180, s=19, v=20)
Color.BURGUNDY = Color(h=150, s=26, v=9)

my_colors = (Color.YELLOW, Color.GREEN, Color.RED, Color.WHITE, Color.GRAY, Color.BURGUNDY, Color.NONE)

sensor.detectable_colors(my_colors)
```

then I have this color detection code:

```python
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

while True:  
    print("looking for green")
    check_for_color(Color.GREEN)

    print("green detected, at station, stopping and continuing")
    print(sensor.hsv())
    motor.brake()
    wait(station_stop_time_ms)
    motor.dc(forward_speed)
```

However, the issue I have is that it's seeing the gray tracks as green. I even tried a lighter/brighter green piece and even recalibrated the HSV value. However, in the print when green is detected, I'm getting values that should be for gray, so I think something isn't right.

Here's my full code in case it helps spot my issue:

```python
from pybricks.pupdevices import ColorSensor, DCMotor # ColorSensor instead of ColorDistanceSensor if using Mindstorms sensor
from pybricks.parameters import Color, Port
from pybricks.tools import wait

check_color_interval_ms = 20
station_stop_time_ms = 1000
eol_stop_time_ms = 1500
forward_speed = 20

motor = DCMotor(Port.A)
sensor = ColorSensor(Port.B) # ColorSensor instead of ColorDistanceSensor if using Mindstorms sensor

Color.YELLOW = Color(h=50, s=20, v=25)
Color.GREEN = Color(h=110, s=30, v=19)
Color.RED = Color(h=354, s=45, v=15)
Color.WHITE = Color(h=195, s=20, v=24)
Color.GRAY = Color(h=180, s=19, v=20)
Color.BURGUNDY = Color(h=150, s=26, v=9)

my_colors = (Color.YELLOW, Color.GREEN, Color.RED, Color.WHITE, Color.GRAY, Color.BURGUNDY, Color.NONE)

sensor.detectable_colors(my_colors)

check_color_interval_ms = 20

def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

while True:  
    print("looking for green")
    check_for_color(Color.GREEN)

    print("green detected, at station, stopping and continuing")
    print(sensor.hsv())
    motor.brake()
    wait(station_stop_time_ms)
    motor.dc(forward_speed)
    
    print("looking for yellow")
    check_for_color(Color.YELLOW)

    print("yellow detected, at end of line, stopping and going back to station")
    print(sensor.hsv())
    motor.brake()
    wait(eol_stop_time_ms)
    motor.dc(-forward_speed)

    print("looking for green")
    check_for_color(Color.GREEN)

    print("green detected, at station, stopping and continuing")
    print(sensor.hsv())
    motor.brake()
    wait(station_stop_time_ms)
    motor.dc(-forward_speed)

    print("looking for red")
    check_for_color(Color.RED)

    print("red detected, at end of line, stopping and going back to station")
    print(sensor.hsv())
    motor.brake()
    wait(eol_stop_time_ms)
    motor.dc(forward_speed)
```

**Comment by dlech** [comment, marked as answer]
Try increasing the saturation (`s=`) values on the colors (but not gray/white). Since they are below 50, it makes the hue component less important in the match. They should be closer to 100 when the hue is important.

**Comment by laurensvalk** [comment]
```python
Color.YELLOW = Color(h=50, s=20, v=25)
Color.GREEN = Color(h=110, s=30, v=19)
Color.RED = Color(h=354, s=45, v=15)
Color.WHITE = Color(h=195, s=20, v=24)
Color.GRAY = Color(h=180, s=19, v=20)
Color.BURGUNDY = Color(h=150, s=26, v=9)
```

These values all look quite dark. I would expect the yellow and green to have much higher values for hue and saturation.

Currently, both yellow/green/white are almost equally unsaturated, so it won't be able to tell them apart too well.

Can you add a picture of your setup? Maybe the sensor is quite far from the surface?

For comparison, see https://www.youtube.com/watch?v=kwnSbnlhOkA

![image](https://github.com/pybricks/support/assets/12326241/ab56ca37-f60c-4f41-b389-e692cc0497cf)

**Comment by Rogue75** [comment]
Thanks for the quick replies dlech & laurensvalk!!

Dlech, I will try turning the saturation up!

It does look like my sensor is quite further than yours, laurensvalk.
![20231215_004651](https://github.com/pybricks/support/assets/7265814/593cf5fc-00c0-485e-a9d3-55c487365d2d)
![20231215_004725](https://github.com/pybricks/support/assets/7265814/5488a907-c9a0-4e16-aeb3-176bbcb599fb)
However, the Eggybrick crew has theirs that recessed. I'll re-run the values for S again, I think I may have lowered them to give it more of a threshold.
<img width="906" alt="eggybrick" src="https://github.com/pybricks/support/assets/7265814/df0cbbf6-5d6a-46e9-bdf8-3a63be4dd39f">

I'll try adjusting values and report back. Amazing software/webware btw!

**Comment by Rogue75** [comment]
I recalibrated Green. I also did Red + Yellow, even tho they work fine.

```python
Color.YELLOW = Color(h=49, s=45, v=27)
Color.GREEN = Color(h=158, s=48, v=17)
Color.RED = Color(h=354, s=50, v=20)
Color.WHITE = Color(h=195, s=20, v=24)
Color.GRAY = Color(h=180, s=19, v=20)
Color.BURGUNDY = Color(h=150, s=26, v=9)
```

and here's what it spit out when hitting the gray railroad piece thinking it was green:

```python
looking for green
green detected, at station, stopping and continuing
Color(h=160, s=24, v=17)
```

If I increase the Green saturation to 100, it runs off the tracks.
If I increase the Green saturation to 75, it still trips on gray:
```python
looking for green
green detected, at station, stopping and continuing
Color(h=180, s=32, v=15)
```
^^^this is the one that bothers me, because I have input nearly the same values for Gray but it sees it as Green...
`Color.GRAY = Color(h=180, s=19, v=20)`

Looks like a value between 75-100 for S on Green was what I needed. I started going up from 75 in increments of 5 and 80 seems to be the magic number!! Thank you!!

Now I need to rewrite the code to treat my Green station as an if statement so that if it doesn't catch green it won't roll off one end of the single track line, but that's for another day. Thanks again.

**Reply by laurensvalk** [reply]
Awesome, can't wait for the video :smile:
